### PR TITLE
FINERACT-1772: Setting Fraud attribute for a loan account

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

in order to mark the loan account as fraud or not. Fineract has a "fraud“ attribute field at the loan account level

[FINERACT-1772](https://issues.apache.org/jira/browse/FINERACT-1772).

Screen Shot 2022-10-19 at 22 57 54

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java [hunk 1]
```diff
@@ -518,6 +518,26 @@ public void validateLoanChargeRefundTransaction(final String json) {
...
+        if (StringUtils.isBlank(json)) {
+            return;
+        }
+        Set<String> transactionParameters = new HashSet<>(Arrays.asList(LoanApiConstants.FRAUD_FLAG_NAME));
+
+        final Type typeOfMap = new TypeToken<Map<String, Object>>() {}.getType();
+        this.fromApiJsonHelper.checkForUnsupportedParameters(typeOfMap, json, transactionParameters);
+
+        final List<ApiParameterError> dataValidationErrors = new ArrayList<>();
+        final DataValidatorBuilder baseDataValidator = new DataValidatorBuilder(dataValidationErrors).resource("loan.flags");
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java]: Extract the "loan.flags" into a constant pls.

Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/serialization/LoanEventApiJsonValidator.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java [hunk 2]
```diff
@@ -994,6 +994,15 @@ public CommandWrapperBuilder updateLoanApplication(final Long loanId) {
         return this;
     }
 
+    public CommandWrapperBuilder markAsFraud(final Long loanId) {
+        this.actionName = "SETFRAUD";
+        this.entityName = "LOAN";
+        this.entityId = loanId;
+        this.loanId = loanId;
+        this.href = "/loans/" + loanId + "/flags";
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java]: Please remove "flags"

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java [hunk 3]
```diff
@@ -161,7 +161,12 @@ public interface LoanApiConstants {
...
     String TRANSACTION_AMOUNT_PARAMNAME = "transactionAmount";
     String PAYMENT_TYPE_PARAMNAME = "paymentTypeId";
+    String FRAUD_FLAG_NAME = "fraud";
 
     // Commands
     String CHARGEBACK_TRANSACTION_COMMAND = "chargeback";
+    String MARK_AS_FRAUD_COMMAND = "markAsFraud";
+
+    // Data Validator names
+    String LOAN_FLAGS_DATAVALIDATOR_PREFIX = "loans.flags";
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java]: Please remove the "flags"

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java [hunk 4]
```diff
@@ -1266,6 +1268,15 @@ private PutLoansLoanIdChanges() {}
         public PutLoansLoanIdChanges changes;
     }
 
+    @Schema(description = "PutLoansLoanIdFlagRequest")
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java]: Its not used anywhere...

Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java]: PutLoansLoanIdRequest can be extended with fraud and be used

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java]: done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java]: done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java (hunk 4 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java]: removed

File: fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java (hunk 2 above)
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/commands/service/CommandWrapperBuilder.java]: please remove the "flags" from the href!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java (hunk 3 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoanApiConstants.java]: nothing changed...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java (hunk 4 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResourceSwagger.java]: nothing changed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java [hunk 5]
```diff
@@ -786,13 +786,22 @@ public String calculateLoanScheduleOrSubmitLoanApplication(
...
+        final CommandWrapperBuilder builder = new CommandWrapperBuilder().withJson(apiRequestBodyAsJson);
+        CommandWrapper commandRequest = null;
+        if (CommandParameterUtil.is(commandParam, LoanApiConstants.MARK_AS_FRAUD_COMMAND)) {
+            commandRequest = builder.markAsFraud(loanId).build();
+        } else {
+            commandRequest = builder.updateLoanApplication(loanId).build();
+        }
 
-        final CommandProcessingResult result = this.commandsSourceWritePlatformService.logCommandSource(commandRequest);
+        if (commandRequest == null) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java]: Is this possible? In the else case above, you always have a commandRequest.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 6]
```diff
@@ -3303,6 +3303,36 @@ public CommandProcessingResult creditBalanceRefund(Long loanId, JsonCommand comm
 
     }
 
+    @Override
+    @Transactional
+    public CommandProcessingResult markLoanAsFraud(Long loanId, JsonCommand command) {
+        Loan loan = this.loanAssembler.assembleFrom(loanId);
+        final Map<String, Object> changes = new LinkedHashMap<>();
+
+        this.loanEventApiJsonValidator.validateMarkAsFraudLoan(command.json());
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: Let's do the validation at the beginning, before assemling the Loan entity. In that case we can save some DB load if there's a validation error.

File: fineract-provider/src/main/resources/db/changelog/tenant/changelog-tenant.xml [hunk 7]
```diff
@@ -81,4 +81,5 @@
     <include file="parts/0059_add_spring_batch_loan_id_list_table.xml" relativeToChangelogFile="true"/>
     <include file="parts/0060_add_job_name_to_command_source.xml" relativeToChangelogFile="true"/>
     <include file="parts/0061_add_idempotency_key_to_command_source.xml" relativeToChangelogFile="true"/>
+    <include file="parts/0062_add_flag_attribute_to_loan.xml" relativeToChangelogFile="true"/>
```
Comment (galovics) [inline fineract-provider/src/main/resources/db/changelog/tenant/changelog-tenant.xml]: let's call it add_fraud_attribute_to_loan.xml

**Review by galovics — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/resources/db/changelog/tenant/changelog-tenant.xml (hunk 7 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/resources/db/changelog/tenant/changelog-tenant.xml]: Done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java (hunk 6 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: Done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java (hunk 5 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java]: Removed

**Comment by adamsaghy** [comment]
LGTM

**Review by galovics — APPROVED** [review]

**Pull request merged**